File: G00376462 - Eoghan Delaney - Assignment 1/python/shop.py
from dataclasses import dataclass, field
from typing import List
import csv

@dataclass
class Product:
    name: str
    price: float = 0.0

@dataclass 
class ProductStock:
    product: Product
    quantity: int

@dataclass 
class Shop:
    cash: float = 0.0
    stock: List[ProductStock] = field(default_factory=list)

@dataclass
class Customer:
    name: str = ""
    budget: float = 0.0
    shopping_list: List[ProductStock] = field(default_factory=list)
    cust_total: float = 0.0

def create_and_stock_shop():
    s = Shop()
    with open('../stock.csv') as csv_file:
        csv_reader = csv.reader(csv_file, delimiter=',')
        first_row = next(csv_reader)
        s.cash = float(first_row[0])
        for row in csv_reader:
            p = Product(row[0], float(row[1]))
            ps = ProductStock(p, float(row[2]))
            s.stock.append(ps)
            #print(ps)
    return s
    
def read_customer(file_path):
    with open(file_path) as csv_file:
        csv_reader = csv.reader(csv_file, delimiter=',')
        first_row = next(csv_reader)
        c = Customer(first_row[0], float(first_row[1]))
        for row in csv_reader:
            name = row[0]
            quantity = float(row[1])
            p = Product(name)
            ps = ProductStock(p, quantity)
            c.shopping_list.append(ps)
        return c 
        

def print_product(p):
    print(f'\nPRODUCT NAME: {p.name} \nPRODUCT PRICE: {p.price}')

def print_customer(c,s):

    print(f"Welcome to the shop {c.name}\nMister money bags! €{c.budget} in your pocket!\nLets see whats on your list?\n")
    cust_total = 0
    for item in c.shopping_list:
        for prod in s.stock:
            if item.product.name==prod.product.name:
                shop_price = prod.product.price
                shop_quant = int(prod.quantity)
        
        cost = item.quantity * shop_price
        if item.quantity >= shop_quant:
            print(f"Sorry, we dont have enough {item.product.name},You want {item.quantity } we only have {shop_quant}, all or nothing!\n")
        else:
            cust_total+=cost
            print(f"NAME OF PRODUCT {item.product.name} PRICE €{shop_price} QUANTITY {item.quantity}\nAmount:€{cost}, SHOP has {shop_quant}\n")
    print(f'Customer Total: €{cust_total}')
    Customer.cust_total = cust_total
    Customer.budget = c.budget
    return Customer
        
def print_shop(s):
    print(f'Shop has {s.cash} in cash')
    for item in s.stock:
        print_product(item.product)
        print(f'The Shop has {item.quantity} of the above')
        
def derping(s, c):
    for item in c.shopping_list:
        for prod in s.stock:
            if item.product.name==prod.product.name:
                print(item)

def find_price(s, c):
    for item in c.shopping_list:
        for prod in s.stock:
            if item.product.name==prod.product.name:
                item.product.price==prod.product.price

def evaluateCustomer(c, s):
    if c.cust_total == 0:
        print("We have nothing you wanted! What a waist of time!\n")
    elif c.cust_total<c.budget:
        print("Looks good\n")
        print(f'Your bill total is €{c.cust_total}, how much did you have again? €{c.budget} \n')
        newshoptotal = s.cash + c.cust_total
        newcusttotal = c.budget-c.cust_total
        print(f"Shop float is now €{newshoptotal}, while {c.name} you now have €{newcusttotal}\n")
        s.cash = newshoptotal
    elif c.cust_total>c.budget:
        print(f'Your bill total is €{c.cust_total}, how much did you have again? €{c.budget} \n')
        print("You don't have enough money! CLEAR OFF!\n")

def askName(s):
    name = input('Whats your name?\n')
    budget = int(input('What is your budget?\n'))
    qtyList = int(input('How many items on your list?\n'))
    print(f'Glad to meet you, Welcome to you and to your €{budget}, {name} - What can i get you?{qtyList} items\n')
    c = Customer(name, budget)
    cust_total = 0
    for _ in range(qtyList):
        prod_name = input('What would you like?\n')
        prod_qty = int(input('How many would you like?\n'))
        p = Product(prod_name)
        ps = ProductStock(p, prod_qty)

        c.shopping_list.append(ps)

        for prod in s.stock:
            if prod_name==prod.product.name:
                shop_price = prod.product.price
                shop_quant = prod.quantity

        
        cost = prod_qty * shop_price
        if prod_qty >= shop_quant:
            print(f"Sorry, we dont have enough {prod_name},You want {prod_qty } we only have {shop_quant}, all or nothing!\n")
        else:
            cust_total+=cost
            print(f"NAME OF PRODUCT {prod_name} PRICE €{shop_price} QUANTITY {prod_qty}\nAmount:€{cost}, SHOP has {shop_quant}\n")
    print(f'Customer Total: €{cust_total}')
    c.cust_total = cust_total
    c.budget = c.budget

    return c



s = create_and_stock_shop()

a = read_customer("../customerB.csv")
b = read_customer("../customerC.csv")
c = read_customer("../customerA.csv")


menu = "************\nPlease choose an option below:\n     1: Customer dose not have enough money.\n     2: Customer is looking for to much of one or more products\n     3: Customer is just right\n     4: Manually enter your order\n     0: To break\n"


def main():
    num = int(input(menu))
    if num == 1:
        answer = print_customer(a, s)
        evaluateCustomer(answer, s)
        main()
    elif num == 2:
        answer = print_customer(b, s)
        evaluateCustomer(answer, s)
        main()
    elif num == 3:
        answer = print_customer(c, s)
        evaluateCustomer(answer, s)
        main()
    elif num == 4:
        Manu_cust = askName(s)
        evaluateCustomer(Manu_cust, s)
        main()
    else:
        print("Bye Bye!")


if __name__ == "__main__":
    main()

File: c/shop.c
#include <stdio.h>
#include <string.h>
#include <stdlib.h>

char *customerA = "../customerA.csv"; // This customer is correct and works out
char *customerB = "../customerB.csv"; // This customer has to little money
char *customerC = "../customerC.csv"; // This customer is looking for to many of one product.



struct Product {
	char* name;
	double price;
};

struct ProductStock {
	struct Product product;
	int quantity;
};

struct Shop {
	double cash;
	struct ProductStock stock[20];
	int index;
};

struct Customer {
	char* name;
	double budget;
	double custotal;
	struct ProductStock shoppingList[10];
	int index;
};

void printProduct(struct Product p)
{
	printf("PRODUCT NAME: %s \nPRODUCT PRICE: %.2f\n", p.name, p.price);
}

double find(struct Shop s,char* name){
	for (int i=0; i<s.index; i++){
		if (strcmp(name, s.stock[i].product.name)==0){
			return s.stock[i].product.price;
		}
	}
	return -1;
}

int findQ(struct Shop s,char* name){
	for (int i=0; i<s.index; i++){
		if (strcmp(name, s.stock[i].product.name)==0){
			return s.stock[i].quantity;
		}
	}
	return -1;
}

void printCustomer(struct Customer c)
{
	printf("CUSTOMER NAME: %s \nCUSTOMER BUDGET: %.2f\n", c.name, c.budget);
	printf("-------------\n");
	for(int i = 0; i < c.index; i++)
	{
		printProduct(c.shoppingList[i].product);	
		printf("%s ORDERS %d OF ABOVE PRODUCT\n", c.name, c.shoppingList[i].quantity);
		double cost = c.shoppingList[i].quantity * c.shoppingList[i].product.price; 
		printf("The cost to %s will be €%.2f\n", c.name, cost);
	}
}

struct Shop createAndStockShop()
{
    FILE * fp;
    char * line = NULL;
    size_t len = 0;
    ssize_t read;

    fp = fopen("../stock.csv", "r");
    if (fp == NULL)
        exit(EXIT_FAILURE);

	read = getline(&line, &len, fp);
	float cash = atof(line);
	// printf("cash in shop is %.2f\n", cash);
	
	struct Shop shop = { cash };

    while ((read = getline(&line, &len, fp)) != -1) {
        // printf("Retrieved line of length %zu:\n", read);
        // printf("%s IS A LINE", line);
		char *n = strtok(line, ",");
		char *p = strtok(NULL, ",");
		char *q = strtok(NULL, ",");
		int quantity = atoi(q);
		double price = atof(p);
		char *name = malloc(sizeof(char) * 50);
		strcpy(name, n);
		struct Product product = { name, price };
		struct ProductStock stockItem = { product, quantity };
		shop.stock[shop.index++] = stockItem;
		// printf("NAME OF PRODUCT %s PRICE %.2f QUANTITY %d\n", name, price, quantity);
    }
	
	return shop;
}

void printShop(struct Shop s)
{
	printf("Shop has %.2f in cash\n", s.cash);
	for (int i = 0; i < s.index; i++)
	{
		printProduct(s.stock[i].product);
		printf("The shop has %d of the above\n", s.stock[i].quantity);
	}
}

struct Customer createAndListCustomer(char *customerSelect){
	FILE * fp;
	char * line = NULL;
	size_t len = 0;
	ssize_t read;

	fp = fopen(customerSelect,"r");
	if (fp == NULL)
		exit(EXIT_FAILURE);

	read = getline(&line,&len, fp);
	// Now split the first line to extract the name and budget.
	char *n = strtok(line, ",");
	char *b = strtok(NULL, ",");
	
	char *name = malloc(sizeof(char) * 50);
	strcpy(name, n);

	double budget= atof(b);
	// struct Customer dominic = { "Eoghan", 100.0};
	
	
	struct Shop shop = createAndStockShop();
	
	char *c ="0.0";
	double custotal= atof(c);
    
	struct Customer customer = { name, budget };

    printf("Welcome to the shop %s\nMister money bags! €%.2f in your pocket!\nLets see whats on your list?\n", n, budget);


	while ((read = getline(&line, &len, fp)) != -1) {
		char *k = strtok(line, ",");
		char *q = strtok(NULL, ",");
			
		int quantity = atoi(q);
		char *name = malloc(sizeof(char) * 50);
		strcpy(name, k);

		double price = find(shop, name);
		int shopQuant = findQ(shop, name);	
		struct Product product = { name };
			
		struct ProductStock shoppingList = { product, quantity };
		customer.shoppingList[customer.index++] = shoppingList;
		
		double amount = price * quantity;
		if (quantity >= shopQuant){
			printf("Sorry, we dont have enough %s,You want %d we only have %d, all or nothing!\n", name, quantity, shopQuant);
		}
		else {
			custotal += amount;
			printf("NAME OF PRODUCT %s PRICE €%.2f QUANTITY %d\nAmount:€%.2f, SHOP has %d\n", name, price, quantity, amount, shopQuant);
		}
	}
	printf("Customer Total: €%.2f\n", custotal);

	customer.custotal = custotal;
	
	return customer;
}

void evaluateCustomer(struct Customer c, struct Shop s){
	if(c.custotal == 0.00){
		printf("We have nothing you wanted! What a waist of time!\n");
	} else if (c.custotal<c.budget){
		printf("Looks good\n");
		printf("Your bill total is €%.2f, how much did you have again? €%.2f \n", c.custotal, c.budget);
		double newtotal = s.cash + c.custotal;
		double newcustotal = c.budget - c.custotal;
		printf("Shop float is now €%.2f, while %s you now have €%.2f\n", newtotal, c.name, newcustotal);
		s.cash = newtotal;
	} else if (c.custotal > c.budget){
		printf("Your bill total is €%.2f, how much did you have again? €%.2f \n", c.custotal, c.budget);
		printf("You dont have enough money! CLEAR OFF!\n");
	};
};


struct Customer askName(){
    char n[10];
	double a;
	int pro;
	char name[50];
	int quantity;
	fflush(stdout);

	printf("What is your name?\n");
	scanf("%s", n);
	printf("What is your budget?\n");
	scanf("%lf", &a);
	struct Customer customer = { n, a };

	printf("How many items on your list?\n");
	scanf("%d", &pro);
	printf("Glad to meet you, Welcome to you and to your €%.2lf, %s - What can i get you?%d items\n", a, n, pro); 
		
	// struct ProductStock shoppingList = { product, quantity };

	
	struct Shop shop = createAndStockShop();
	
	char *c ="0.0";
	double custotal= atof(c);

	for(int i = 0; i < pro; i++)
	{
		printf("What would you like and how many?%d\n", i);
		scanf("%5s %d", name, &quantity);
		fflush(stdout);

		//int quantity = atoi(q);
		//char *name = malloc(sizeof(char) * 50);
		//strcpy(name, k);
		
		double price = find(shop, name);
		int shopQuant = findQ(shop, name);
		struct Product product = { name };

		struct ProductStock shoppingList = { product, quantity };
		customer.shoppingList[customer.index++] = shoppingList;
		
		double amount = price * quantity;
		if (quantity >= shopQuant){
			printf("Sorry, we dont have enough %s,You want %d we only have %d, all or nothing!\n", name, quantity, shopQuant);
		}
		else {
			custotal += amount;
			printf("NAME OF PRODUCT %s PRICE €%.2f QUANTITY %d\nAmount:€%.2f, SHOP has %d\n", name, price, quantity, amount, shopQuant);
		}
	}
	printf("Customer Total: €%.2f\n", custotal);

	customer.custotal = custotal;
	
	return customer;
}

int main(void) 
{	
	int choice =-1;

	struct Shop shop = createAndStockShop();
	//printShop(shop);


	while(choice !=0 ){
		fflush(stdin);
		printf("************\nPlease choose an option below:\n    1: Customer dose not have enough money.\n    2: Customer is looking for to much of one or more products\n    3: Customer is just right\n    4: Manually enter your order\n    0: To break\n");
		scanf("%d",&choice);

		if(choice == 1){
			struct Customer customer = createAndListCustomer(customerB);
			evaluateCustomer(customer, shop);
		}
		else if (choice == 2){
			struct Customer customer = createAndListCustomer(customerC);
			evaluateCustomer(customer, shop);
		}
		else if (choice == 3){
			struct Customer customer = createAndListCustomer(customerA);
			evaluateCustomer(customer, shop);
		}
		else if (choice == 4){
			struct Customer customer = askName();
			//printShop(shop);
			evaluateCustomer(customer, shop);
		}
		else if (choice == 0){
			printf("Bye");
			return 0;
		}
	}
	printf("Bye");
	
	//printf("***** LETS HAVE A LOOK AT THE SHOP ***********\n");
	//struct Shop shop = createAndStockShop();
	//printShop(shop);
	
	//printf("***** CAN I HELP YOU? ***********\n");
	//struct Customer customer = createAndListCustomer(customerB);
	
	//printf("***** REVIEW THE ORDER ***********\n");
	//evaluateCustomer(customer, shop);
    return 0;
}